program/src/state/mod.rs

```diff
@@ -42,45 +42,52 @@ pub enum Key {
     Frequency,
 }
 
-pub trait SolanaAccount: BorshDeserialize {
+pub trait SolanaAccount: BorshSerialize + BorshDeserialize {
+    /// Get the `Key` for this `Account`.  This key is to be stored in the first byte of the
+    /// `Account` data.
     fn key() -> Key;
 
+    /// Get the size of the `Account` data.
     fn size() -> usize;
 
-    fn is_correct_account_type(data: &[u8], data_type: Key) -> bool {
-        let key: Option<Key> = Key::from_u8(data[0]);
-        match key {
-            Some(key) => key == data_type || key == Key::Uninitialized,
-            None => false,
-        }
-    }
-
-    fn pad_length(buf: &mut Vec<u8>) -> Result<(), RuleSetError> {
-        let padding_length = Self::size()
-            .checked_sub(buf.len())
-            .ok_or(RuleSetError::NumericalOverflow)?;
-        buf.extend(vec![0; padding_length]);
-        Ok(())
-    }
+    /// BorshDeserialize the `AccountInfo` into the Rust data structure.
+    fn from_account_info(account: &AccountInfo) -> Result<Self, ProgramError> {
+        let data = account
+            .data
+            .try_borrow()
+            .map_err(|_| ProgramError::AccountBorrowFailed)?;
 
-    fn safe_deserialize(mut data: &[u8]) -> Result<Self, BorshError> {
-        if !Self::is_correct_account_type(data, Self::key()) {
-            return Err(BorshError::new(ErrorKind::Other, "DataTypeMismatch"));
+        if !Self::is_correct_account_type_and_size(&data, Self::key(), Self::size()) {
```

Comment:
I'm not crazy about the size check because it's not really checkable for things like dynamically sized accounts.

Reply:
I'll remove the size check, I also don't like it for that reason.